qubes/tests/integ/vm_qrexec_gui.py

```diff
@@ -65,6 +65,276 @@ def setUp(self):
+            "while ! pactl list sink-inputs | grep -q :{}; do sleep 1; done'"
+            .format(local_user, vm.name)))
+        self.loop.run_until_complete(p.wait())
+        # and some more...
+        self.loop.run_until_complete(asyncio.sleep(1))
+
+    def prepare_audio_vm(self, backend):
+        if 'whonix-gw' in self.template:
+            self.skipTest('whonix-gw have no audio')
+        self.loop.run_until_complete(self.testvm1.start())
+        pulseaudio_units = 'pulseaudio.socket pulseaudio.service'
+        pipewire_units = 'pipewire.socket wireplumber.service pipewire.service'
+        if backend == 'pipewire':
+            if not self.testvm1.features.check_with_template('supported-service.pipewire', False):
+                self.skipTest('PipeWire not supported in VM')
+            if 'debian' in self.template or 'whonix' in self.template:
+                self.skipTest('PipeWire audio not supported in Debian')
+            self.testvm1.features['service.pipewire'] = True
+        elif backend == 'pulseaudio':
+            # Use PulseAudio if it is installed.  If it is not installed,
+            # PipeWire will still run, and its PulseAudio emulation will
+            # be tested.
+            self.testvm1.features['service.pipewire'] = False
+        else:
+            self.fail('bad audio backend')
+        self.wait_for_pulseaudio_startup(self.testvm1)
+
+    def check_audio_sample(self, sample, sfreq):
+        rec = np.fromstring(sample, dtype=np.float32)
+        # determine sample size using silence threshold
+        threshold = 10**-3
+        rec_size = np.count_nonzero((rec > threshold) | (rec < -threshold))
+        if not rec_size:
+            self.fail('only silence detected, no useful audio data')
+        # find zero crossings
+        crossings = np.nonzero((rec[1:] > threshold) &
+                            (rec[:-1] < -threshold))[0]
+        np.seterr('raise')
+        # compare against sine wave frequency
+        rec_freq = 44100/np.mean(np.diff(crossings))
+        if not sfreq*0.8 < rec_freq < sfreq*1.2:
+            fname = f"/tmp/audio-sample-{self.id()}.raw"
+            with open(fname, "wb") as f:
+                f.write(sample)
+            self.fail('frequency {} not in specified range, saved to {}'
+                    .format(rec_freq, fname))
+
+    def common_audio_playback(self):
+        # sine frequency
+        sfreq = 4400
+        # generate signal
+        audio_in = np.sin(2*np.pi*np.arange(44100)*sfreq/44100)
+        # Need to use .snd extension so that pw-play (really libsndfile)
+        # recognizes the file as raw audio.
+        self.loop.run_until_complete(
+            self.testvm1.run_for_stdio('cat > audio_in.snd',
+            input=audio_in.astype(np.float32).tobytes()))
+        local_user = grp.getgrnam('qubes').gr_mem[0]
+        if self.testvm1.features['service.pipewire']:
+            cmd = 'pw-play --format=f32 --rate=44100 --channels=1 audio_in.snd'
```

Comment:
Here is the fix:
```suggestion
            cmd = 'pw-play --format=f32 --rate=44100 --channels=1 - < audio_in.snd'
```
 :)